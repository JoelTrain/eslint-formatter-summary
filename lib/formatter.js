import chalk from 'chalk';
import parseArgs from './parse-cli-args';
import pad from './pad-num';
import sum from './sum-up';
import lengthOfLongest from './length-of-longest';
import sortBy from './sort-by-prop';
import aggregate from './aggregator';
import processMessage from './process-message';

function constructHeader(len) {
  const sparkles = String.fromCodePoint(0x2728);
  const header = ` ${sparkles}  Summary of failing ESLint rules`.padEnd(len);
  return chalk`{bgBlue ${header}}`;
}

function constructSummary(rules) {
  const errorLen = lengthOfLongest('errors', rules);
  const warningLen = lengthOfLongest('warnings', rules);
  return rules.map(function print(rule) {
    const errors = pad(rule.errors, errorLen);
    const warnings = pad(rule.warnings, warningLen);
    const line = chalk`errors {red ${errors}} warnings {yellow ${warnings}} rule: {gray ${rule.ruleId}}`;
    return `${line}\n`;
  }).join('');
}

function constructTotal(summary) {
  const errors = sum('errors', summary);
  const warnings = sum('warnings', summary);
  const problems = errors + warnings;
  const flames = String.fromCodePoint(0x1f525);
  return chalk`${flames}  {red ${problems} problems in total} (${errors} {red errors}, ${warnings} {yellow warnings})`;
}

export function format(results, processArgv) {
  const rules = aggregate(results);

  const args = parseArgs(processArgv);
  const direction = args.desc ? 'desc' : 'asc';

  if (args.sortBy === 'rule') {
    sortBy('ruleId', rules, direction);
  }
  if (args.sortBy === 'errors') {
    sortBy('errors', rules, direction);
  }
  if (args.sortBy === 'warnings') {
    sortBy('warnings', rules, direction);
  }

  const errorLen = lengthOfLongest('errors', rules);
  const warningLen = lengthOfLongest('warnings', rules);
  const ruleIdLen = lengthOfLongest('ruleId', rules);
  const headerLen = `errors ${'0'.repeat(errorLen)} warnings ${'0'.repeat(warningLen)} rule: ${'0'.repeat(ruleIdLen)}`.length;

  const header = constructHeader(headerLen);
  const summary = constructSummary(rules);
  const total = constructTotal(rules);

  return `${header}\n${summary}\n${total}`;
}

function formatter(results) {
  console.log(format(results, process.argv));
};

export default formatter;
