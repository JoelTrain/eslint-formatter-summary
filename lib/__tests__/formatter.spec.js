import { format } from '../formatter';
import mockResult from './mock-result-factory';

jest.unmock('yargs');

test('provides header, summary of rules and total', () => {
  const results = [
    mockResult([[ 'rule1', 1 ], [ 'rule2', 1 ]]),
    mockResult([[ 'rule2', 1 ], [ 'rule3', 2 ]]),
    mockResult([[ 'rule3', 2 ], [ 'rule4', 2 ], [ 'rule5', 2 ]])
  ];
  const argv = [];
  expect(format(results, argv)).toMatchSnapshot();
});

test('can sort summary by rule id ascending', () => {
  const results = [
    mockResult([[ 'rule2', 1 ], [ 'rule3', 2 ]]),
    mockResult([[ 'rule1', 1 ], [ 'rule2', 1 ]])
  ];
  const argv = [
    'eslint',
    '-f',
    './index.js',
    '.',
    '--',
    '--sort-by',
    'rule'
  ];
  expect(format(results, argv)).toMatchSnapshot();
});

test('can sort summary by rule id descending', () => {
  const results = [
    mockResult([[ 'rule2', 1 ], [ 'rule3', 2 ]]),
    mockResult([[ 'rule1', 1 ], [ 'rule2', 1 ]])
  ];
  const argv = [
    'eslint',
    '-f',
    './index.js',
    '.',
    '--',
    '--sort-by',
    'rule',
    '--desc'
  ];
  expect(format(results, argv)).toMatchSnapshot();
});

test('can sort summary by num of errors ascending', () => {
  const results = [
    mockResult([[ 'rule2', 2 ], [ 'rule3', 2 ]]),
    mockResult([[ 'rule1', 1 ], [ 'rule2', 2 ]])
  ];
  const argv = [
    'eslint',
    '-f',
    './index.js',
    '.',
    '--',
    '--sort-by',
    'errors'
  ];
  expect(format(results, argv)).toMatchSnapshot();
});

test('can sort summary by num of errors descending', () => {
  const results = [
    mockResult([[ 'rule2', 2 ], [ 'rule3', 2 ]]),
    mockResult([[ 'rule1', 1 ], [ 'rule2', 2 ]])
  ];
  const argv = [
    'eslint',
    '-f',
    './index.js',
    '.',
    '--',
    '--sort-by',
    'errors',
    '--desc'
  ];
  expect(format(results, argv)).toMatchSnapshot();
});
