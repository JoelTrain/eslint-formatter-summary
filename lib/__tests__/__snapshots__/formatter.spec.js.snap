// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can sort summary by rule id ascending 1`] = `
"[0m[44m âœ¨  Summary of failing ESLint rules[49m[0m
[0merrors [0m[0m[31m0[39m[0m[0m warnings [0m[0m[33m1[39m[0m[0m rule: [0m[0m[90mrule1[39m[0m
[0merrors [0m[0m[31m0[39m[0m[0m warnings [0m[0m[33m2[39m[0m[0m rule: [0m[0m[90mrule2[39m[0m
[0merrors [0m[0m[31m1[39m[0m[0m warnings [0m[0m[33m0[39m[0m[0m rule: [0m[0m[90mrule3[39m[0m

[0mðŸ”¥  [0m[0m[31m4 problems in total[39m[0m[0m (1 [0m[0m[31merrors[39m[0m[0m, 3 [0m[0m[33mwarnings[39m[0m[0m)[0m"
`;

exports[`can sort summary by rule id descending 1`] = `
"[0m[44m âœ¨  Summary of failing ESLint rules[49m[0m
[0merrors [0m[0m[31m1[39m[0m[0m warnings [0m[0m[33m0[39m[0m[0m rule: [0m[0m[90mrule3[39m[0m
[0merrors [0m[0m[31m0[39m[0m[0m warnings [0m[0m[33m2[39m[0m[0m rule: [0m[0m[90mrule2[39m[0m
[0merrors [0m[0m[31m0[39m[0m[0m warnings [0m[0m[33m1[39m[0m[0m rule: [0m[0m[90mrule1[39m[0m

[0mðŸ”¥  [0m[0m[31m4 problems in total[39m[0m[0m (1 [0m[0m[31merrors[39m[0m[0m, 3 [0m[0m[33mwarnings[39m[0m[0m)[0m"
`;

exports[`provides header, summary of rules and total 1`] = `
"[0m[44m âœ¨  Summary of failing ESLint rules[49m[0m
[0merrors [0m[0m[31m0[39m[0m[0m warnings [0m[0m[33m1[39m[0m[0m rule: [0m[0m[90mrule1[39m[0m
[0merrors [0m[0m[31m0[39m[0m[0m warnings [0m[0m[33m2[39m[0m[0m rule: [0m[0m[90mrule2[39m[0m
[0merrors [0m[0m[31m2[39m[0m[0m warnings [0m[0m[33m0[39m[0m[0m rule: [0m[0m[90mrule3[39m[0m
[0merrors [0m[0m[31m1[39m[0m[0m warnings [0m[0m[33m0[39m[0m[0m rule: [0m[0m[90mrule4[39m[0m
[0merrors [0m[0m[31m1[39m[0m[0m warnings [0m[0m[33m0[39m[0m[0m rule: [0m[0m[90mrule5[39m[0m

[0mðŸ”¥  [0m[0m[31m7 problems in total[39m[0m[0m (4 [0m[0m[31merrors[39m[0m[0m, 3 [0m[0m[33mwarnings[39m[0m[0m)[0m"
`;
